### PR TITLE
[CR] Standardize overmap terrain identifiers

### PR DESCRIPTION
**Pull request by codemime**
This PR, to some extent, follows #16646 and #17450. 
Will conflict with #18367.

Main issue with the old identifiers (apart from duplication and broken encapsulation) is that they heavily use string comparison instead of numbers, which slows down already not-so-fast overmap generation/exploration. 

No observable changes meant.
### Implementation details
- `oter_special` map (unused) was removed
- implicit conversions to strings/objects (error prone) were removed as well
- `string_id<>` and `int_id<>` were reused to replace the old identifiers
- overmap terrain data structures (`oterlist` and `otermap`) are local now (aren't exported from the header)

EDIT:
- `int_id<>` was extended to support `->` syntax (like `string_id<>` already does).

I stop here to keep the reasonable PR size. In a following PR I'm going to reuse `generic_factory` and noticeably reduce the number of string comparisons.

### CODE REVIEW

File: src/omdata.h [hunk 1]
```diff
@@ -119,7 +111,14 @@ class oter_id
         size_t size() const;
 };
 
+/// @todo: Deprecate these operators
```
Comment (mugling) [inline src/omdata.h]: I don't think you need both `///` and `@todo`. Presuming that's correct the latter is clearer

File: src/bionics.cpp [hunk 2]
```diff
@@ -467,7 +467,7 @@ bool player::activate_bionic( int b, bool eff_only )
             vehwindspeed = abs(veh->velocity / 100); // vehicle velocity in mph
         }
         const oter_id &cur_om_ter = overmap_buffer.ter( global_omt_location() );
-        std::string omtername = otermap[cur_om_ter].name;
+        const std::string &omtername = cur_om_ter.obj().name;
```
Comment (mugling) [inline src/bionics.cpp]: `string_id` allows indirection of the form `cur_om_ter->name`. You could extend `int_id` to include a similar `operator *` to avoid having to write `obj()` everywhere

File: src/editmap.cpp [hunk 3]
```diff
@@ -1836,8 +1835,8 @@ int editmap::mapgen_preview( real_coords &tc, uimenu &gmenu )
 
                 } else if( gpmenu.ret == 3 ) {
                     popup( _( "Changed oter_id from '%s' (%s) to '%s' (%s)" ),
-                           orig_oters.t().name.c_str(), orig_oters.c_str(),
-                           omt_ref.t().name.c_str(), omt_ref.c_str() );
+                           orig_oters.obj().name.c_str(), orig_oters.id().c_str(),
```
Comment (mugling) [inline src/editmap.cpp]: This idiom crops up a lot. It might be worth extending `int_id` to provide a native optional `c_str()` method

File: src/map.h [hunk 4]
```diff
@@ -40,8 +40,11 @@ class computer;
 struct itype;
 struct mapgendata;
 struct trap;
+struct oter_t;
+
 using trap_id = int_id<trap>;
-struct oter_id;
+using oter_id = int_id<oter_t>;
```
Comment (mugling) [inline src/map.h]: These aren't nice names. I appreciate that wasn't invented here. How badly would line count be affected by more descriptive naming?

File: src/omdata.h [hunk 5]
```diff
@@ -45,9 +46,13 @@ enum oter_flags {
...
 };
 
+struct oter_t;
+using oter_id = int_id<oter_t>;
+using oter_str_id = string_id<oter_t>;
+
 struct oter_t {
-        std::string id;      // definitive identifier
-        unsigned loadid;          // position in termap / terlist
+        oter_str_id id;      // definitive identifier
```
Comment (mugling) [inline src/omdata.h]: New comments could be doxygen style

File: src/bionics.cpp (hunk 2 above)
Comment (codemime) [inline src/bionics.cpp]: Good idea.

File: src/editmap.cpp (hunk 3 above)
Comment (codemime) [inline src/editmap.cpp]: It would be slightly inconsistent if you don't rename `id()` to `str()` as well.

File: src/map.h (hunk 4 above)
Comment (codemime) [inline src/map.h]: > These aren't nice names.

It's not an ideal name, but it's perfectly consistent with other ids. What would you suggest instead?

> How badly would line count be affected by more descriptive naming?

Massively. I've counted 623 occurrences. I definitely don't want to to it here.

File: src/editmap.cpp (hunk 3 above)
Comment (mugling) [inline src/editmap.cpp]: True. How many lines would that change touch?

File: src/map.h (hunk 4 above)
Comment (mugling) [inline src/map.h]: To stay as it is then for now

**Comment by mugling** [comment]
> Will conflict with #18367 (that's why it's marked WIP).

It might be easier to try for a merge of this PR first. Given you've marked it `CR` I'll wait until either @Coolthulhu, @kevingranade or @BevapDin have taken a look first

**Comment by codemime** [comment]
> It might be easier to try for a merge of this PR first.

I definitely want it to be merged soon (I expect it can easily accumulates merge conflicts otherwise) and I don't really care which one goes first. So I don't mind.

**Comment by mugling** [comment]
> I definitely want it to be merged soon (I expect it can easily accumulates merge conflicts otherwise) and I don't really care which one goes fir

It's reasonable to expect priority here. Unpleasant work on evil code shouldn't be made any harder than necessary.

File: src/editmap.cpp (hunk 3 above)
Comment (codemime) [inline src/editmap.cpp]: Not so many (~80), but in various parts of the code, so I think it deserves a separate PR.

**Comment by codemime** [comment]
Thanks.

**Comment by mugling** [comment]
If you grok the the map code it would be great to see someone improve mapgen. Natural obstacles like large deserts, cliffs or especially large bodies of water water would be epic

**Comment by codemime** [comment]
> If you grok the the map code it would be great to see someone improve mapgen. Natural obstacles like large deserts, cliffs or especially large bodies of water water would be epic

I like the idea and thought about it too, but the Augean stables should be cleaned first.

**Comment by codemime** [comment]
All set.

**Comment by mugling** [comment]
Needs agreement as to direction from another developer, if so will start merge test

**Comment by Rivet-the-Zombie** [comment]
Looks good to me.
